### PR TITLE
GLFW 3.3

### PR DESCRIPTION
**Pull request by pchampio**
This PR adds support for the latest stable version of GLFW ([3.3-stable from github](https://github.com/glfw/glfw/commit/e4e9581557826bd522158ab9159859e20e10700e)). 

I've addressed the issues of #219 :
* [x] idiomatic Go function (returns `(*Thing, error)`) 
* [x] Check for new methods implemented on GLFW side since  #219 was made. (http://www.glfw.org/docs/3.3/news.html)
* [x] Check issue list for fixes postponed until 3.3 (except #235)
* [x] Re-base
* [ ] Make sure vulkan's working. I've ommited Vulkan specific things but we've had reports that they are necessary. Last time I've checked vulkan-go's glfw fork was not so different.

  _(Not realy sure what to do here)_ Even if vulkan isn't very polished, important bug fixes, and other features are available.

I've renamed the v3.2 folder to v3.3 and then done the changes, this allows a 'cleaner' PR. [I can also make this even cleaner with some filter (removes first commit)](https://github.com/go-gl/glfw/pull/256/files/4a65e39dc111391137546213600348a5bed9bff2..ecfa2137ee25dae782dee47463fe31724fd72df9)
I'll add the v3.2 folder back after everything is green on you side.

### CODE REVIEW

**Comment by pchampio** [comment]
I'm pretty sure vulkan isn't working on darwin.
We need to set some compile flag for this platform.
This can be done in another PR!!!

I am runing my tests using one of 'my' project (https://github.com/go-flutter-desktop/hover | https://github.com/go-flutter-desktop/go-flutter).
hover allows for cross compiling, and I can assert that each platform does compile properly.

Here are an example of hover created application using go-gl/glfw v3.3: 
windows: [windows.zip](https://github.com/go-gl/glfw/files/3800101/windows.zip)
linux: [linux.zip](https://github.com/go-gl/glfw/files/3800100/linux.zip)
macos:  [darwin.zip](https://we.tl/t-dvyibVFW3T) (7 days only, wetransfer link)

**Comment by pchampio** [comment]
A tag named 'v3.3/glfw/v0.1.0' needs to be pushed (and updated) for go modules to work.

**Comment by xlab** [comment]
> I'm pretty sure vulkan isn't working on darwin.
> We need to set some compile flag for this platform.

Hey, any specific error report for that? I'm pretty sure that for darwin it requires just one more framework `-framework Metal -framework MoltenVK`
https://github.com/vulkan-go/glfw/blob/v3.3/v3.3/glfw/build.go#L19

Is there something beyond that?

**Comment by xlab** [comment]
Amazing work, I hope this will be merged soon. I especially like the concept of having Vulkan in another PR. Let's merge GLFW mainline, then we'll consider how to handle vulkan part in separate PR.

**Comment by pchampio** [comment]
> Hey, any specific error report for that? I'm pretty sure that for darwin it requires just one more framework -framework Metal -framework MoltenVK
https://github.com/vulkan-go/glfw/blob/v3.3/v3.3/glfw/build.go#L19
> 
> Is there something beyond that?

If I put those flags, then cross-compiling using Docker does not work anymore. 

---

Keep in mind that those LD-flag should only be used when the builg tag 'vulkan' is set (like the README says)

```golang
/*
#cgo darwin,vulkan CFLAGS: -D_GLFW_VULKAN_STATIC
#cgo darwin,vulkan LDFLAGS: -F/Library/Frameworks -framework Metal -framework MoltenVK -framework IOSurface -framework QuartzCore
*/
import "C"
```
TODO (In another PR) port back vulkan functions from https://github.com/vulkan-go/glfw/commit/2a913f74e92e1c26522a1bffa497a56f3b08fc18

**Comment by kivutar** [comment]
I'm not sure about the `#include "glfw/src/linux_joystick.c"` in the freebsd build. It gives me errors. Is it something new?

**Comment by pchampio** [comment]
What's your error?
And no it's not something new, the linux_joystick import hasn't changed, it was already here in GLFW 3.2.

If you look at the GLFW imports [here](https://github.com/glfw/glfw/tree/3.3-stable/src) the only available files are {cocoa,linux,null,win32}_joystick.*
Freebsd should use the Linux import.

**Comment by kivutar** [comment]
My error is about some missing symbols because of a missing linux specific header. I can't copy the error messages because I'm using VirtualBox.

These lines also suggest that this file is Linux specific: https://github.com/go-gl/glfw/pull/256/files#diff-e4ef75560f509962482af9c7dc7397e5R66-R69

But you're right that this is nothing new, and we may just ignore it for now.

**Comment by rcoreilly** [comment]
What is the status of 3.3 here?  I see this PR, and a 3.3-beta branch which just received a commit -- are there two separate 3.3 versions?  One way or another, it would be great to get 3.3 finally available after a really long time since the upstream lib was released.  Thanks!

**Comment by pchampio** [comment]
Your comment is misleading. 
The 3.3-beta branch (#219) hasn't received a any commit since last year. https://github.com/go-gl/glfw/pull/219/commits

As discussed on https://github.com/go-gl/glfw/pull/219#issuecomment-485657318 and stated in my PR description, I'm taken over #219.

I haven't received any updates from the maintainers of go-gl. I've invested not payed time into this PR (like everyone😑), and I also would like it get merged before upstream lib diverge too much.

**Comment by rcoreilly** [comment]
Sorry I was confused -- I received an email notification about #257 which looked new but apparently it was just closed! And I mistook the name of that for the 3.3-beta branch.. Anyway, I guess the only valid remaining point is: when will this PR be merged!?

**Comment by pchampio** [comment]
I can see why you got confused, I realllllly hope this PR gets merged before the end of the year.
If not, even though I don't like the idea, I will certainly start my own fork 😒.

(I don't want to throw the work I've done, and I don't want to stress out open source maintainers, as I know how hard it can be) 

Speaking of which, if you want to add me as a contributor for go-gl/glfw, I will be honored!

**Comment by kivutar** [comment]
@pchampio you may want to take a look at these lines for OSX https://github.com/go-gl/glfw/pull/219/files#diff-c114ff13290eefc1d62b13112e94f1e5R13-R19

I don't know what `-D_GLFW_USE_RETINA` does exactly, but maybe it would help some people.

You can certainly add `-framework Metal` safely.

However, since not everybody has `-framework MoltenVK`, adding this framework would better be done with a build flag. If not it would break `go get` for users that don't have MoltenVK installed.

I can help you test this branch on OSX at least for OpenGL, as I'm already maintaining my own fork for 3.3. I would prefer to not use a fork.

**Comment by hajimehoshi** [comment]
> However, since not everybody has -framework MoltenVK, adding this framework would better be done with a build flag. If not it would break go get for users that don't have MoltenVK installed.

Agreed. `MoltenVK` should be optional.

**Comment by pchampio** [comment]
> @pchampio you may want to take a look at these lines for OSX https://github.com/go-gl/glfw/pull/219/files#diff-c114ff13290eefc1d62b13112e94f1e5R13-R19
> 
> I don't know what `-D_GLFW_USE_RETINA` does exactly, but maybe it would help some people.

I've intentionally removed those line (as described in https://github.com/go-gl/glfw/pull/256/commits/a643862e73a009e920f3d6ad735788c4858f90d8 commit :smiley:) because those compile-time macros have been removed in 3.3. https://www.glfw.org/docs/3.3/news.html#removals_33

> You can certainly add `-framework Metal` safely.

It breaks compatibility with MacOSX10.10.sdk.
I'm basing this conclusion on https://github.com/docker/golang-cross.
> https://s3.dockerproject.org/darwin/v2/MacOSX10.10.sdk.tar.xz **Doesn't not contain** `Metal.framework`
> while https://raw.githubusercontent.com/ndeloof/golang-cross/113fix/MacOSX10.11.sdk.tar.xz does.

I don't think `-framework Metal` is necessary.  https://github.com/glfw/glfw/blob/2bac7ee8da526257d808bd026b027246c98e4f2f/CMakeLists.txt#L277-L297
Am I missing something?

**Comment by frou** [comment]
All Macs that can run 10.10 can also run 10.11, so in theory no one is stuck on 10.10

**Comment by xlab** [comment]
Afaik Vulkan works through Metal, so afaik it's only needed for MoltenVK to work.

**Comment by pchampio** [comment]
Sooo.., Is the PR ready to be reviewed?
Friendly ping @tapir, @pwaller, @dmitshur, @slimsag.

File: v3.3/glfw/GLFW_C_REVISION.txt
Comment (pwaller) [inline v3.3/glfw/GLFW_C_REVISION.txt:1]: I note that this choice of revision is after the `3.3` tag.

Is that really what we want? Or do we want to stick to "released" versions? I could be persuaded either way.

**Review by pwaller — APPROVED** [review]
@pchampio thanks very much for this effort, and everyone else who chimed in.

I'm low on time, but I have done my best to move all the bits through my eyeballs. It looks good to me, and thanks for your efforts on making it easy to review. I only found a couple of trivial things to comment on.

Moving forward, my intent would be to merge this with the "squash" feature on github so that we don't get the removal/addition of v3.2 in the git history.

I will give a couple of days in case anyone else has any comments, but unless there is a show stopper, I don't see any reason not to merge. I agree with the rationale "Vulkan may not be working, but this is a step forward". Future PRs can fix other things.

If there is no further input on this thread by Saturday 23rd Nov, then I will merge. If I don't merge then, ping the thread and I'll do it by the end of the day on Sunday -- and if I don't respond then, I am inundated, so please feel free to ping the thread weekly or so.

File: README.md
Comment (pchampio) [reply README.md]: Moved into the `XX specific changes` header

File: v3.3/glfw/window.go
Comment (pchampio) [reply v3.3/glfw/window.go]: reversed the changes, I tried another implementation but it didn't workout. Forgot to reverse the godoc.

File: v3.3/glfw/GLFW_C_REVISION.txt
Comment (pchampio) [reply v3.3/glfw/GLFW_C_REVISION.txt:1]: On the [glfw.org/download](https://www.glfw.org/download.html) webpage, it is stated:
>  the 3.3-stable branch only adds bug fixes for patch releases.

I'm guessing we want the latest patch available in `go-gl/glfw`

**Comment by pwaller** [comment]
Merged. Great work, thanks!

**Comment by pchampio** [comment]
Just for the sake of sharing numbers.
After a very fast investigation, here are some well stared project that rely on go-gl/glfw:
* https://github.com/go-flutter-desktop/go-flutter 3.3k :+1: 
* https://github.com/faiface/pixel 2.6k :+1: 
* https://github.com/hajimehoshi/ebiten 2.2k :+1: 
* https://github.com/goki/gi  570 :+1: 
* https://github.com/libretro/ludo 230 :+1:

Thanks to the go-gl team :heart: !
Putting the 5-6h of work into this PR was well worth-it!

**Comment by hajimehoshi** [comment]
hajimehoshi/ebiten has already switched to 3.3. @pchampio, thank you for your awesome work!

**Comment by dmitshur** [comment]
Thank you @pchampio for authoring this PR, @pwaller for reviewing it, @tapir for creating the original PR #219, and everyone else who contributed to this change! 🎉 ❤️ 🚀

I've sent [CL 208877](https://golang.org/cl/208877) to update [`golang.org/x/exp/shiny/driver/mtldriver`](https://pkg.go.dev/golang.org/x/exp/shiny/driver/mtldriver) to use the new GLFW v3.3 library.

**Comment by kivutar** [comment]
libretro/ludo is also switching to 3.3. Thanks to all the people involved in this PR.

**Comment by rcoreilly** [comment]
Yep same with goki/gi and I can already say that it has made things more stable and functional!  Thank you!

**Pull request merged**
